### ISSUE TITLE
Refer to Vega’s JSON schema for parts that are similar between Vega-lite and Vega

### ISSUE THREAD

**Issue by kanitw** [vega/vega-lite#620]
This is now possible given that Vega 2 also has a schema.

**Comment by kanitw** [comment]
Also need to follow  [these instructions](https://github.com/vega/vega/wiki/Vega-and-Browserify) to resolve the missing package errors

**Comment by kanitw** [comment]
@arvind  I try using vega’s schema but some of the `require` is broken in browserify-compiled version.
(Running it in the unit test works fine.)  

For example,  the following line results in `undefined` in the compiled version in vega-lite editor and gallery.  

```js
var vgStackSchema = require('vega/src/transforms/Stack').schema;
```

I guess this is due to the fact that `module.exports = Stack;` is called before `Stack.schema` is defined in `Stack.js`
I am saying this because I tried moving `module.exports = Stack;` to the end of line in `Stack.js` in my `node_modules/vega` and no longer get errors.   

(Yeah it’s quite weird — I thought it does not matter where we put module.exports in node.js)

**Comment by arvind** [comment]
Yes, there is a browserify transform that strips out the schema definitions from the exported modules and puts them into a separate `vega-schema.json` file. The transform assumes that anything after  `module.exports` is a schema definition and strips it out. This allows us to keep the built distributions of Vega compact.

**Comment by arvind** [comment]
The browserify transform is certainly non-standard for node (`module.exports` should be valid anywhere). Ultimately, I would like to co-locate the schema definitions and code so I don't mind moving to another delimiter to trigger the extraction. 

But that doesn't seem to be at the core of your question which is that you're trying to validate against a specific section of the Vega schema. I can restructure the `vega-schema.json` to make each of the transforms addressable via JSONref rather than having them being nested as `oneOf` within `#/defs/transforms` as they are currently?

**Comment by kanitw** [comment]
> you're trying to validate against a specific section of the Vega schema

What I’m doing here is to define an enum in vega-lite schema by referring to Vega’s schema.  

See the commented part in 104e93fa58054b6e23ec4e6eb1eaace92b94fb1c for example.

**Comment by kanitw** [comment]
From offline conversation, @arvind  suggests that we use reference and compile the schema, so we do not have to package the whole vega codebase.  

He will also refactor the Vega schema to have transforms properly namespaced (rather than having them all in `oneOf`.)

**Comment by arvind** [comment]
Ok, with vega/vega@5d9ae05, you should now be able to address individual transform schemas. E.g., `defs.aggregateTransform` or `defs.stackTransform`. You'll need to perform a custom build (`npm run schema`) until we release the next version of Vega. Hope this helps!

**Comment by kanitw** [comment]
@arvind  Thanks a lot.  This is definitely helpful.  I guess Vega should be our "devDependency" and I will need to use `npm link` so that the custom build is referred in node?

One quick question -- How do you normally decide to release new version of Vega?  A schema structure change likes this sounds like a legitimate PATCH or even MINOR version for me.  
(Just curious -- no pressure on doing this now as this is more like a refactor issue than high priority issue for us.)

**Comment by arvind** [comment]
We will be releasing a patch version soon, hopefully within the week. There's another bug that Jane filed that I want to take a look at once I have the time.

Regarding versioning, we've tried to hew to semantic versioning as much as possible. Sometimes, however, we have pushed changes more aggressively than the version numbers suggest. The tension with version numbers is that not only do they have a functional component (semantic versioning denoting the state of public APIs) but also a marketing component. So we try and balance between the two. 

For example, while the structure of the schema can technically be construed as a public API, I imagine that the vast majority of use-cases aren't addressing specific subsections of the schema directly. And, even if they are, the above change extends the API rather than introduces a breaking change. As a result, I am more comfortable including it within a patch release rather than a minor release.
